### DISCUSSION TITLE
Systemd not enabled

### DISCUSSION THREAD

**Discussion by SSP6904** [community/community#68954] [Codespaces]
### Select Topic Area

Question

### Body

I'm using codespace for using with apache. However, systemd and systemctl are all not working because of it's overhead.

How come it's doing this error?

```
"systemd" is not running in this container due to its overhead.
Use the "service" command to start services instead. e.g.: 
```

**Comment by RandomGamingIn144p** [comment]
same for me lol

**Reply by beoceafy-inc** [reply]
same bro

**Comment by Bossa9973** [comment]
I need support on that too

**Comment by karl6986** [comment]
me too

**Comment by bidianqing** [comment]
me too

**Comment by mrlanwr** [comment]
me too

**Comment by tesla59** [comment]
Codespaces run inside containers and it is not possible to run systemd inside containers
ref: https://github.com/orgs/community/discussions/54655#discussioncomment-7338124

**Reply by gbraad** [reply]
I run systemd in containers since ever.

```dockerfile
ENTRYPOINT ["/sbin/init"]
```

and you can also use `docker` to build them and use:

```dockerfile
RUN systemctl enable httpd
```

... ~but `docker` is limited to not run systemd correctly, while `podman` can~.

> [!NOTE]
> the reference you gave indicates the limitation "run as a Docker container"; that's the problem, as it is possible with almost any other container runtime. Correction: Docker can and will run `systemd`, and also works for Codespaces, just not `container:` on Actions.

**Reply by gbraad** [reply]
> [!IMPORTANT]
> I got this working as follows: https://github.com/gbraad-devenv/fedora-systemd/

I put a short summarized answer here: https://github.com/orgs/community/discussions/68954#discussioncomment-11812864

**Comment by gbraad** [comment]
I got this working as follows. It is two commands that are essential:

`.devcontainer.json`
```json
    "overrideCommand": false,
    "privileged": true
```

and making sure:

`Dockerfile`
```Dockerfile
USER root
ENTRYPOINT [ "/sbin/init" ]
```

is set for the container.

I made an example here: https://github.com/gbraad-devenv/fedora-systemd/

#### Try it out

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/gbraad-devenv/fedora-systemd)

> [!NOTE]
> This uses a personalized Fedora-based image that I published on [`gbraad-devenv`](https://github.com/gbraad-devenv/fedora/), but this is also possible with the 'stock' images on Codespaces; just create a `Dockerfile` with `FROM mcr.microsoft.com/devcontainers/base:ubuntu`. More information here: https://hub.docker.com/r/microsoft/devcontainers-base. You might have to install the `systemd`-package, as I have not tested these. I do not use someone else's :toothbrush: .

**Reply by gbraad** [reply]
I eventually added a default `.devcontainer` using the Microsoft provided Ubuntu base:
 - https://github.com/gbraad-dotfiles/upstream/blob/main/.devcontainer/default/Containerfile
 - https://github.com/gbraad-dotfiles/upstream/blob/main/.devcontainer/default/devcontainer.json

Mostly reused what I had before, so it might contain some redundant packages.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/gbraad-dotfiles/upstream)

![image](https://github.com/user-attachments/assets/f7946a61-79fb-4aa1-9998-2aba4da1cdee)

It shows degraded, as the getty@tty1 service fails. Kinda expected and not a big deal.

![image](https://github.com/user-attachments/assets/2749c5ff-fb5c-48dd-af11-0c26188da90d)
